File: solutions/0823. Binary Trees With Factors/0823.py
# Time complexity: O(n^2)
# Space complexity: O(n)
class Solution:
  def numFactoredBinaryTrees(self, arr: List[int]) -> int:
    kMod = 1_000_000_007
    n = len(arr)
    dp = [1] * n
    arr.sort()
    numToIndex = {a: i for i, a in enumerate(arr)}

    for i, root in enumerate(arr):
      for j in range(i):
        if root % arr[j] == 0:
          right = root // arr[j]
          if right in numToIndex:
            dp[i] += dp[j] * dp[numToIndex[right]]
            dp[i] %= kMod

    return sum(dp) % kMod        

File: solutions/0823. Binary Trees With Factors/0823.rs
// Time complexity: O(n^2)
// Space complexity: O(n)
use std::collections::HashMap;

impl Solution {
    pub fn num_factored_binary_trees(mut arr: Vec<i32>) -> i32 {
        let k_mod = 1_000_000_007;
        let mut dp = vec![1; arr.len()];
        let mut num_to_index: HashMap<i32, usize> = HashMap::new();

        arr.sort_unstable();

        for (i, &root) in arr.iter().enumerate() {
            for j in 0..i {
                if root % arr[j] == 0 {
                    let right = root / arr[j];
                    if let Some(&index) = num_to_index.get(&right) {
                        dp[i] += dp[j] * dp[index];
                        dp[i] %= k_mod;
                    }
                }
            }
            num_to_index.insert(root, i);
        }

        (dp.iter().fold(0, |sum, &e| sum + e as i64) % k_mod) as i32
    }
}
